Практическое задание № 5. 

Шаг 1. Загрузка данных
1.1. Используйте набор данных, содержащий последовательные данные (например, временные ряды, текстовые последовательности или любые другие данные, содержащие последовательную структуру).

In [1]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import SimpleRNN, Dense
from tensorflow.keras.optimizers import Adam
import numpy as np

In [2]:
# Шаг 1: Загрузка данных
# Пример: последовательность чисел от 1 до 100
data = np.arange(1, 101, 1)

1.2. Загрузите данные и подготовьте их для использования в модели RNN.

In [4]:
# Создаем последовательные наборы данных для ввода и вывода RNN
input_sequence = [data[i:i+5] for i in range(len(data)-5)]
output_sequence = [data[i+5] for i in range(len(data)-5)]

In [5]:
# Преобразуем данные в массивы NumPy
X = np.array(input_sequence).reshape(-1, 5, 1)
y = np.array(output_sequence)

Шаг 2. Создание рекуррентной нейронной сети (RNN)

2.1. Импортируйте необходимые библиотеки, такие как TensorFlow и Keras.

In [6]:
model = Sequential()
model.add(SimpleRNN(50, activation='relu', input_shape=(5, 1)))
model.add(Dense(1))

Шаг 3. Обучение RNN

In [7]:
model.compile(optimizer=Adam(), loss='mean_squared_error')
model.fit(X, y, epochs=50, batch_size=1)


Epoch 1/50

95/95 [==============================] - 7s 5ms/step - loss: 583.5588
Epoch 2/50
95/95 [==============================] - 1s 7ms/step - loss: 1.3941
Epoch 3/50
95/95 [==============================] - 1s 7ms/step - loss: 1.3471
Epoch 4/50
95/95 [==============================] - 1s 5ms/step - loss: 1.2088
Epoch 5/50
95/95 [==============================] - 1s 6ms/step - loss: 1.1814
Epoch 6/50
95/95 [==============================] - 1s 5ms/step - loss: 1.0462
Epoch 7/50
95/95 [==============================] - 1s 5ms/step - loss: 0.9930
Epoch 8/50
95/95 [==============================] - 1s 5ms/step - loss: 0.9370
Epoch 9/50
95/95 [==============================] - 0s 5ms/step - loss: 0.8891
Epoch 10/50
95/95 [==============================] - 1s 6ms/step - loss: 0.7511
Epoch 11/50
95/95 [==============================] - 1s 6ms/step - loss: 0.7880
Epoch 12/50
95/95 [==============================] - 1s 8ms/step - loss: 0.7214
Epoch 13/50
95/95 [===========================

Шаг 4. Оценка производительности модели RNN

In [8]:
test_data = np.arange(101, 111, 1)
X_test = np.array([test_data[i:i+5] for i in range(len(test_data)-5)])
y_test = np.array([test_data[i+5] for i in range(len(test_data)-5)])


In [9]:
# Предсказание на тестовых данных
predictions = model.predict(X_test.reshape(-1, 5, 1))

# Оценка производительности
mse = np.mean((predictions.flatten() - y_test) ** 2)
print("Среднеквадратичная ошибка на тестовых данных:", mse)

1/1 [==============================] - 1s 874ms/step
Среднеквадратичная ошибка на тестовых данных: 0.4387305185664445


Значение MSE равное 0.4387 на тестовых данных говорит о том, что модель довольно точно предсказывает следующее число в последовательности. Однако, так как это относительная метрика, её интерпретация может зависеть от контекста поставленной задачи.